Attempting to resolve #5

agharbeia opened the pull request:
This is an attempt to resolve issue #5 , by adding a field to the configuration form of the module to hold the internal name of the HumHub profile attribute where ORCIDs should be stored.

Upon saving the form, if no value is entered in the abovementioned field, "orcid" is assumed.

A better solution would be to do this during the validation of the form, and leave the configuration saving procedure unmodified.

Other modifications: Added GPL license file, fixed module metadata, and added codemeta.json as per CodeMeta project.

ArchBlood committed:
Merge pull request #2 from GreenMeteor/develop

Fix: scopes

agharbeia committed:
Added the GPL license as per our agreement.
Please refer to my email message dated 2024-12-13 for reference
regarding the choice of license.

agharbeia committed:
Attempting to resolve #5 :
Pre-fill ORCID in log-in form, if entered in HumHub profile

agharbeia committed:
Fixed URLs and email address.
Added A. Gharbeia as contributor

agharbeia committed:
Added metadata as per the CodeMeta project

ArchBlood commented:
Please update https://github.com/GreenMeteor/auth-orcid/blob/main/docs/LICENCE.md linking to the license;

### Example
- https://www.gnu.org/licenses/gpl-3.0.en.html

Currently HumHub's marketplace only renders md file such as `LICENSE.md`, this makes it easier.

ArchBlood commented on module.json:
```diff
@@ -12,10 +12,17 @@
+            "name": "Ahmad Gharbeia",
+            "email": "ahmad@gharbeia.org",
+            "homepage": "https://ahmad@gharbeia.org",
             "role": "Developer"
```
Please change role to `"Collaborator"` or `"Contributor"`

agharbeia replied:
Sure. That was my intention. I didn't know the possible keywords, though, or if it were a controlled vocabulary.

ArchBlood commented on codemeta.json:
```diff
@@ -0,0 +1,46 @@
+{
```
If we plan on adding `codemeta.json` then we will also need to include a `.gitignore` for packaged releases. 🤔
https://docs.github.com/en/get-started/git-basics/ignoring-files

agharbeia replied:
Is this something I can do on my side?

ArchBlood replied:
I've changed the branch to develop which includes this already.

ArchBlood commented on models/ConfigureForm.php:
```diff
@@ -52,6 +53,15 @@ public function saveSettings()
         $settings->set('clientId', $this->clientId);
         $settings->set('clientSecret', $this->clientSecret);
+        
+        /* On saving, if no value is entered for orcidAttribute, then assign "orcid" to it.
```
Use `//` comments please as using `/* */` is normally seen as disabling "dead code".

agharbeia committed:
Ahmad as contributor.

agharbeia committed:
Linking to the GPL3 license

agharbeia commented:
> Please update https://github.com/GreenMeteor/auth-orcid/blob/main/docs/LICENCE.md linking to the license;
> Currently HumHub's marketplace only renders md file such as `LICENSE.md`, this makes it easier.

Can't we do both, @ArchBlood?

1. The GPL license requires the inclusion of its text with the source code
2. GitHub can only identify the licence of the code and label the repo as being licensed as such, when the text of the licence is included in root in a file named LICENCE.

ArchBlood commented:
> > Please update https://github.com/GreenMeteor/auth-orcid/blob/main/docs/LICENCE.md linking to the license;
> > Currently HumHub's marketplace only renders md file such as `LICENSE.md`, this makes it easier.
> 
> Can't we do both, @ArchBlood?
> 
> 1. The GPL license requires the inclusion of its text with the source code
> 2. GitHub can only identify the licence of the code and label the repo as being licensed as such, when the text of the licence is included in root in a file named LICENCE.

Just linking should satisfy this so it is still compliant.

agharbeia committed:
Made configuration fields clientId' and  'clientSecret' required.
fixed comment style.

agharbeia committed:
Fixed typo in in-string variable name.

agharbeia committed:
Fixed typo in name of in-string variable 'adminLink'.

ArchBlood commented on codemeta.json:
```diff
@@ -0,0 +1,46 @@
+{
```
I'm still on the fence for this, maybe use something like this? But overall I'd prefer a GitHub action workflow that would handle generation and updating of this file instead of manual. 🤔
```json
{
  "@context": "https://doi.org/10.5063/schema/codemeta-2.0",
  "@type": "SoftwareSourceCode",
  "@id": "https://github.com/GreenMeteor/auth-orcid",
  "name": "ORCID Authentication",
  "description": "HumHub module for OAuth via ORCID.org",
  "codeRepository": "https://github.com/GreenMeteor/auth-orcid.git",
  "issueTracker": "https://github.com/GreenMeteor/auth-orcid/issues",
  "version": "1.0.0",
  "dateCreated": "2024-12-13",
  "dateModified": "2025-03-29",
  "datePublished": "2024-12-13",
  "programmingLanguage": "PHP",
  "softwareRequirements": "https://github.com/humhub/humhub",
  "license": "https://spdx.org/licenses/LGPL-3.0+",
  "author": [
    {
      "@id": "https://github.com/ArchBlood",
      "@type": "Person",
      "givenName": "ArchBlood",
      "email": "kodo@greenmeteor.net",
      "affiliation": {
        "@type": "Organization",
        "name": "GreenMeteor"
      }
    }
  ],
  "contributor": [
    {
      "@id": "https://orcid.org/0000-0003-3930-7811",
      "@type": "Person",
      "givenName": "Ahmad",
      "familyName": "Gharbeia",
      "email": "ahmad@gharbeia.org"
    }
  ],
  "keywords": ["HumHub", "ORCID", "Authentication", "Module", "OAuth"]
}
```

ArchBlood commented on authclient/OrcidClient.php:
```diff
@@ -189,11 +190,11 @@ public function afterSave($user, $attributes)
 
         if (!empty($attributes['id'])) {
-            $orcidField = ProfileField::findOne(['internal_name' => 'orcid']);
+            $fieldName = $this->orcidAttribute
```
Missing `;` here.

ArchBlood commented on authclient/OrcidClient.php:
```diff
@@ -189,11 +190,11 @@ public function afterSave($user, $attributes)
         if (!empty($attributes['id'])) {
-            $orcidField = ProfileField::findOne(['internal_name' => 'orcid']);
+            $fieldName = $this->orcidAttribute
+            $orcidField = ProfileField::findOne(['internal_name' => $this->$fieldName]);
```
This is also incorrect and should be `$fieldName` not `$this->$fieldName`.

ArchBlood commented on models/ConfigureForm.php:
```diff
@@ -34,6 +34,7 @@ public function attributeLabels()
         return [
             'enabled' => Yii::t('AuthOrcidModule.base', 'Enabled'),
             'clientId' => Yii::t('AuthOrcidModule.base', 'ORCID Client ID'),
+            'orcidAttribute' => Yii::t('AuthOrcidModule.base', 'Internal name of profile attribute to hold ORCIDs. If not entered, {orcidAttribute} will be used.', ['orcidAttribute' => Html::tag('code', 'orcid')]),
```
I'm not sure that `attributeLabels()` would allow for this as it would render as plaintext, also you're applying a Html class without a use statement which would throw an error.

```txt
Internal name of profile attribute to hold ORCIDs. If not entered, <code>orcid</code> will be used.
```

Just use `['orcidAttribute' => 'orcid']` and it should be fine

ArchBlood commented on authclient/OrcidClient.php:
```diff
@@ -57,6 +57,7 @@ public function init()
             $this->scope = $config->scopes;
             $this->clientId = $config->clientId;
             $this->clientSecret = $config->clientSecret;
+            $this->orcidAttribute = $config->orcidAttribute;
```
Where there is no `public $orcidAttribute` this is not possible to be used it should either be removed or defined properly before use.

ArchBlood commented on models/ConfigureForm.php:
```diff
@@ -52,6 +53,15 @@ public function saveSettings()
+        // Ideally, this should take place during form validation, by setting a default value
+        // to the form-field, and then letting the saving of configuration to proceed as usual.
+        if ($this->orcidAttribute == null) {
+        	$this->orcidAttribute = 'orcid';
```
Has no valid `public $oridAttribute` needs corrected.

ArchBlood reviewed (changes requested):
Thank you very much for this good addition to the module, a few things to take into account here.

ArchBlood committed:
Update ConfigureForm.php

Signed-off-by: ArchBlood <35392110+ArchBlood@users.noreply.github.com>

ArchBlood committed:
Refactor OrcidClient for improved readability

Signed-off-by: ArchBlood <35392110+ArchBlood@users.noreply.github.com>

ArchBlood committed:
Revise codemeta.json for improved metadata

Updated codemeta.json to include new metadata fields and corrected author information.

Signed-off-by: ArchBlood <35392110+ArchBlood@users.noreply.github.com>

ArchBlood committed:
Update module.json

Signed-off-by: ArchBlood <35392110+ArchBlood@users.noreply.github.com>
